File: plug-ins/dev/python/n_gfUtilParentConstraint.py
# -*- coding: utf-8 -*-
"""
Copyright (c) 2019 Giuliano França

MIT License

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

====================================================================================================

Disclaimer:
    THIS PLUGIN IS JUST A PROTOTYPE. YOU MUST USE THE C++ RELEASE PLUGIN FOR PRODUCTION.
    YOU CAN FIND THE C++ RELEASE PLUGIN FOR YOUR SPECIFIC PLATFORM IN RELEASES FOLDER:
    "gfTools > plug-ins > release"

How to use:
    * Copy the parent folder to the MAYA_SCRIPT_PATH.
    * To find MAYA_SCRIPT_PATH paste this command in a Python tab:
        import os; os.environ["MAYA_SCRIPT_PATH"].split(";")
    * In Maya, go to Windows > Settings/Preferences > Plug-in Manager.
    * Browse for "gfTools > plug-ins > dev > python"
    * Find gfTools_P.py and import it.

Requirements:
    * Maya 2017 or above.

Description:
    Custom parent contraint. Parent one object to another. Can be used as point, orient
    or scale constraint.

Attributes:
    * Offset: The matrix offset between the source and target objects.
    * Target World Matrix: The world matrix of the target object.
    * Constraint Parent Inverse Matrix: The world inverse matrix of the parent of the constrainted object.
    * Out Constraint: The output matrix of the constraint.

Todo:
    * Add support to joint orient.
    * Add support to rotation order.
    * Add support to weights.

Sources:
    * NDA

This code supports Pylint. Rc file in project.
"""

import maya.api._OpenMaya_py2 as om2


def maya_useNewAPI():
    """ Function to Maya recognize the use of the Python API 2.0. """
    # pylint: disable=invalid-name, unnecessary-pass
    pass


def INPUT_ATTR(FNATTR):
    """ Configure a input attribute. """
    # pylint: disable=invalid-name
    FNATTR.writable = True
    FNATTR.readable = True
    FNATTR.storable = True
    FNATTR.keyable = True


def OUTPUT_ATTR(FNATTR):
    """ Configure a output attribute. """
    # pylint: disable=invalid-name
    FNATTR.writable = False
    FNATTR.readable = True
    FNATTR.storable = False
    FNATTR.keyable = False


class ParentConstraint(om2.MPxNode):
    """ Main class of gfUtilParentConstraint node. """

    kNodeName = ""
    kNodeClassify = ""
    kNodeID = ""

    inConstraintJntOri = om2.MObject()
    inConstraintRotOrder = om2.MObject()
    inConstraintParInvMtx = om2.MObject()
    inConstraintParSca = om2.MObject()
    # inTargetJntOri = om2.MObject()
    # inTargetRotOrder = om2.MObject()
    inTargetWorldMatrix = om2.MObject()
    inTargetOffset = om2.MObject()
    inTargetWeight = om2.MObject()
    inTargetList = om2.MObject()
    outConstTrans = om2.MObject()
    outConstRot = om2.MObject()
    outConstSca = om2.MObject()

    def __init__(self):
        """ Constructor. """
        om2.MPxNode.__init__(self)

    @staticmethod
    def creator():
        """ Maya creator function. """
        return ParentConstraint()

    @staticmethod
    def initialize():
        """
        Defines the set of attributes for this node. The attributes declared in this function are assigned
        as static members to ParentConstraint class. Instances of ParentConstraint will use these attributes to create plugs
        for use in the compute() method.
        """
        mAttr = om2.MFnMatrixAttribute()
        nAttr = om2.MFnNumericAttribute()
        uAttr = om2.MFnUnitAttribute()
        eAttr = om2.MFnEnumAttribute()
        cAttr = om2.MFnCompoundAttribute()

        constJntOriX = uAttr.create("constraintJointOrientX", "cjorx", om2.MFnUnitAttribute.kAngle, 0.0)
        constJntOriY = uAttr.create("constraintJointOrientY", "cjory", om2.MFnUnitAttribute.kAngle, 0.0)
        constJntOriZ = uAttr.create("constraintJointOrientZ", "cjorz", om2.MFnUnitAttribute.kAngle, 0.0)
        ParentConstraint.inConstraintJntOri = nAttr.create("constraintJointOrient", "cjor", constJntOriX, constJntOriY, constJntOriZ)
        INPUT_ATTR(nAttr)

        ParentConstraint.inConstraintRotOrder = eAttr.create("constraintRotateOrder", "croo", 0)
        eAttr.addField("xyz", 0)
        eAttr.addField("yzx", 1)
        eAttr.addField("zxy", 2)
        eAttr.addField("xzy", 3)
        eAttr.addField("yxz", 4)
        eAttr.addField("zyx", 5)
        INPUT_ATTR(eAttr)

        ParentConstraint.inConstraintParInvMtx = mAttr.create("constraintParentInverseMatrix", "cpim", om2.MFnMatrixAttribute.kDouble)
        INPUT_ATTR(mAttr)

        ParentConstraint.inConstraintParSca = nAttr.createPoint("constraintParentScale", "cps")
        nAttr.default = (1.0, 1.0, 1.0)
        INPUT_ATTR(nAttr)

        ParentConstraint.inTargetWorldMatrix = mAttr.create("targetWorldMatrix", "twmtx", om2.MFnMatrixAttribute.kDouble)
        INPUT_ATTR(mAttr)

        ParentConstraint.inTargetOffset = mAttr.create("targetOffset", "toff", om2.MFnMatrixAttribute.kDouble)
        INPUT_ATTR(mAttr)

        # targetJntOriX = uAttr.create("targetJointOrientX", "tjorx", om2.MFnUnitAttribute.kAngle, 0.0)
        # targetJntOriY = uAttr.create("targetJointOrientY", "tjory", om2.MFnUnitAttribute.kAngle, 0.0)
        # targetJntOriZ = uAttr.create("targetJointOrientZ", "tjorz", om2.MFnUnitAttribute.kAngle, 0.0)
        # ParentConstraint.inTargetJntOri = nAttr.create("targetJointOrient", "tjor", targetJntOriX, targetJntOriY, targetJntOriZ)
        # INPUT_ATTR(nAttr)

        # ParentConstraint.inTargetRotOrder = eAttr.create("targetRotateOrder", "troo", 0)
        # eAttr.addField("xyz", 0)
        # eAttr.addField("yzx", 1)
        # eAttr.addField("zxy", 2)
        # eAttr.addField("xzy", 3)
        # eAttr.addField("yxz", 4)
        # eAttr.addField("zyx", 5)
        # INPUT_ATTR(eAttr)

        ParentConstraint.inTargetWeight = nAttr.create("targetWeight", "twght", om2.MFnNumericData.kFloat, 1.0)
        INPUT_ATTR(nAttr)

        ParentConstraint.inTargetList = cAttr.create("targetList", "tlist")
        cAttr.addChild(ParentConstraint.inTargetWorldMatrix)
        cAttr.addChild(ParentConstraint.inTargetOffset)
        # cAttr.addChild(ParentConstraint.inTargetJntOri)
        # cAttr.addChild(ParentConstraint.inTargetRotOrder)
        cAttr.addChild(ParentConstraint.inTargetWeight)
        cAttr.array = True

        ParentConstraint.outConstTrans = nAttr.createPoint("constraintTranslate", "ctrans")
        OUTPUT_ATTR(nAttr)

        outConstRotX = uAttr.create("constraintRotateX", "crox", om2.MFnUnitAttribute.kAngle, 0.0)
        outConstRotY = uAttr.create("constraintRotateY", "croy", om2.MFnUnitAttribute.kAngle, 0.0)
        outConstRotZ = uAttr.create("constraintRotateZ", "croz", om2.MFnUnitAttribute.kAngle, 0.0)
        ParentConstraint.outConstRot = nAttr.create("constraintRotate", "cro", outConstRotX, outConstRotY, outConstRotZ)
        OUTPUT_ATTR(nAttr)

        ParentConstraint.outConstSca = nAttr.createPoint("constraintScale", "csca")
        nAttr.default = (1.0, 1.0, 1.0)
        OUTPUT_ATTR(nAttr)

        ParentConstraint.addAttribute(ParentConstraint.inConstraintJntOri)
        ParentConstraint.addAttribute(ParentConstraint.inConstraintRotOrder)
        ParentConstraint.addAttribute(ParentConstraint.inConstraintParInvMtx)
        ParentConstraint.addAttribute(ParentConstraint.inConstraintParSca)
        ParentConstraint.addAttribute(ParentConstraint.inTargetList)
        ParentConstraint.addAttribute(ParentConstraint.outConstTrans)
        ParentConstraint.addAttribute(ParentConstraint.outConstRot)
        ParentConstraint.addAttribute(ParentConstraint.outConstSca)
        ParentConstraint.attributeAffects(ParentConstraint.inConstraintJntOri, ParentConstraint.outConstTrans)
        ParentConstraint.attributeAffects(ParentConstraint.inConstraintRotOrder, ParentConstraint.outConstTrans)
        ParentConstraint.attributeAffects(ParentConstraint.inConstraintParInvMtx, ParentConstraint.outConstTrans)
        ParentConstraint.attributeAffects(ParentConstraint.inConstraintParSca, ParentConstraint.outConstTrans)
        ParentConstraint.attributeAffects(ParentConstraint.inTargetWorldMatrix, ParentConstraint.outConstTrans)
        ParentConstraint.attributeAffects(ParentConstraint.inTargetOffset, ParentConstraint.outConstTrans)
        # ParentConstraint.attributeAffects(ParentConstraint.inTargetJntOri, ParentConstraint.outConstTrans)
        # ParentConstraint.attributeAffects(ParentConstraint.inTargetRotOrder, ParentConstraint.outConstTrans)
        ParentConstraint.attributeAffects(ParentConstraint.inTargetWeight, ParentConstraint.outConstTrans)
        ParentConstraint.attributeAffects(ParentConstraint.inConstraintJntOri, ParentConstraint.outConstRot)
        ParentConstraint.attributeAffects(ParentConstraint.inConstraintRotOrder, ParentConstraint.outConstRot)
        ParentConstraint.attributeAffects(ParentConstraint.inConstraintParInvMtx, ParentConstraint.outConstRot)
        ParentConstraint.attributeAffects(ParentConstraint.inConstraintParSca, ParentConstraint.outConstRot)
        ParentConstraint.attributeAffects(ParentConstraint.inTargetWorldMatrix, ParentConstraint.outConstRot)
        ParentConstraint.attributeAffects(ParentConstraint.inTargetOffset, ParentConstraint.outConstRot)
        # ParentConstraint.attributeAffects(ParentConstraint.inTargetJntOri, ParentConstraint.outConstRot)
        # ParentConstraint.attributeAffects(ParentConstraint.inTargetRotOrder, ParentConstraint.outConstRot)
        ParentConstraint.attributeAffects(ParentConstraint.inTargetWeight, ParentConstraint.outConstRot)
        ParentConstraint.attributeAffects(ParentConstraint.inConstraintJntOri, ParentConstraint.outConstSca)
        ParentConstraint.attributeAffects(ParentConstraint.inConstraintRotOrder, ParentConstraint.outConstSca)
        ParentConstraint.attributeAffects(ParentConstraint.inConstraintParInvMtx, ParentConstraint.outConstSca)
        ParentConstraint.attributeAffects(ParentConstraint.inConstraintParSca, ParentConstraint.outConstSca)
        ParentConstraint.attributeAffects(ParentConstraint.inTargetWorldMatrix, ParentConstraint.outConstSca)
        ParentConstraint.attributeAffects(ParentConstraint.inTargetOffset, ParentConstraint.outConstSca)
        # ParentConstraint.attributeAffects(ParentConstraint.inTargetJntOri, ParentConstraint.outConstSca)
        # ParentConstraint.attributeAffects(ParentConstraint.inTargetRotOrder, ParentConstraint.outConstSca)
        ParentConstraint.attributeAffects(ParentConstraint.inTargetWeight, ParentConstraint.outConstSca)

    def compute(self, plug, dataBlock):
        """
        Node computation method:
            * plug is a connection point related to one of our node attributes (either an input or an output).
            * dataBlock contains the data on which we will base our computations.
        """
        # pylint: disable=no-self-use
        eConstJntOri = om2.MEulerRotation(dataBlock.inputValue(ParentConstraint.inConstraintJntOri).asDouble3())
        mConstParInv = dataBlock.inputValue(ParentConstraint.inConstraintParInvMtx).asMatrix()
        constRotOrder = dataBlock.inputValue(ParentConstraint.inConstraintRotOrder).asShort()
        constParSca = dataBlock.inputValue(ParentConstraint.inConstraintParSca).asFloat3()
        targetListHandle = dataBlock.inputArrayValue(ParentConstraint.inTargetList)
        mTargetsAdded = om2.MMatrix()
        mtxFn = om2.MTransformationMatrix()
        mtxFn.scaleBy(constParSca, om2.MSpace.kTransform)
        mInvSca = mtxFn.asMatrix()

        for i in range(len(targetListHandle)):
            targetListHandle.jumpToLogicalElement(i)
            targetHandle = targetListHandle.inputValue()
            # targetJntOri = om2.MEulerRotation(targetHandle.child(ParentConstraint.inTargetJntOri).asDouble3())
            # targetRotOrder = targetHandle.child(ParentConstraint.inTargetRotOrder).asShort()
            mTargetW = targetHandle.child(ParentConstraint.inTargetWorldMatrix).asMatrix()
            mOffset = targetHandle.child(ParentConstraint.inTargetOffset).asMatrix()
            targetWeight = targetHandle.child(ParentConstraint.inTargetWeight).asFloat()

            mTarget = mOffset * (mTargetW * targetWeight)
            if mTargetsAdded == om2.MMatrix():
                mTargetsAdded = mTarget
            else:
                mTargetsAdded += mTarget

        mResult = mTargetsAdded * mConstParInv * mInvSca

        if plug == ParentConstraint.outConstTrans:
            outTransHandle = dataBlock.outputValue(ParentConstraint.outConstTrans)
            outTrans = om2.MFloatVector(mResult[12], mResult[13], mResult[14])
            outTransHandle.setMFloatVector(outTrans)
            outTransHandle.setClean()
        if plug == ParentConstraint.outConstRot:
            outRotHandle = dataBlock.outputValue(ParentConstraint.outConstRot)
            mtxFn = om2.MTransformationMatrix(mResult)
            eRotMtx = mtxFn.rotation(asQuaternion=False)
            qRotMtx = eRotMtx.asQuaternion()
            qConstJntOri = eConstJntOri.asQuaternion()
            qOutRot = qRotMtx * qConstJntOri.invertIt()
            outRot = qOutRot.asEulerRotation().reorderIt(constRotOrder)
            outRotHandle.setMVector(outRot.asVector())
            outRotHandle.setClean()
        if plug == ParentConstraint.outConstSca:
            outScaHandle = dataBlock.outputValue(ParentConstraint.outConstSca)
            mtxFn = om2.MTransformationMatrix(mResult)
            outSca = mtxFn.scale(om2.MSpace.kWorld)
            outScaHandle.set3Float(outSca[0], outSca[1], outSca[2])
            outScaHandle.setClean()

File: plug-ins/dev/cpp/source/n_gfUtilParentConstraint.cpp
#include "headers/n_gfUtilParentConstraint.h"

// Configure a input attribute.
#define INPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(true);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(true);	\
	FNATTR.setKeyable(true);	\

// Configure a output attribute.
#define OUTPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(false);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(false);	\
	FNATTR.setKeyable(false);	\


// Constructor.
ParentConstraint::ParentConstraint() {}

// Destructor.
ParentConstraint::~ParentConstraint() {}

MObject ParentConstraint::inConstraintJntOri;
MObject ParentConstraint::inConstraintRotOrder;
MObject ParentConstraint::inConstraintParInvMtx;
MObject ParentConstraint::inConstraintParSca;
MObject ParentConstraint::inTargetWorldMatrix;
MObject ParentConstraint::inTargetOffset;
MObject ParentConstraint::inTargetWeight;
MObject ParentConstraint::inTargetList;
MObject ParentConstraint::outConstTrans;
MObject ParentConstraint::outConstRot;
MObject ParentConstraint::outConstSca;


void* ParentConstraint::creator(){
    // Maya creator function.
    return new ParentConstraint();
}

MStatus ParentConstraint::initialize(){
    /*
    Defines the set of attributes for this node. The attributes declared in this function are assigned
    as static members to ParentConstraint class. Instances of ParentConstraint will use these attributes to create plugs
    for use in the compute() method.
    */
    MStatus status;
    MFnMatrixAttribute mAttr;
    MFnNumericAttribute nAttr;
    MFnUnitAttribute uAttr;
    MFnEnumAttribute eAttr;
    MFnCompoundAttribute cAttr;

    MObject constJntOriX = uAttr.create("constraintJointOrientX", "cjorx", MFnUnitAttribute::kAngle, 0.0, &status);
    MObject constJntOriY = uAttr.create("constraintJointOrientY", "cjory", MFnUnitAttribute::kAngle, 0.0, &status);
    MObject constJntOriZ = uAttr.create("constraintJointOrientZ", "cjorz", MFnUnitAttribute::kAngle, 0.0, &status);
    inConstraintJntOri = nAttr.create("constraintJointOrient", "cjor", constJntOriX, constJntOriY, constJntOriZ, &status);
    INPUT_ATTR(nAttr);

    inConstraintRotOrder = eAttr.create("constraintRotateOrder", "croo", 0, &status);
    eAttr.addField("xyz", 0);
    eAttr.addField("yzx", 1);
    eAttr.addField("zxy", 2);
    eAttr.addField("xzy", 3);
    eAttr.addField("yxz", 4);
    eAttr.addField("zyx", 5);
    INPUT_ATTR(eAttr);

    inConstraintParInvMtx = mAttr.create("constraintParentInverseMatrix", "cpim", MFnMatrixAttribute::kDouble, &status);
    INPUT_ATTR(mAttr);

    inConstraintParSca = nAttr.createPoint("constraintParentScale", "cps", &status);
    nAttr.setDefault(1.0, 1.0, 1.0);
    INPUT_ATTR(nAttr);

    inTargetWorldMatrix = mAttr.create("targetWorldMatrix", "twmtx", MFnMatrixAttribute::kDouble, &status);
    INPUT_ATTR(mAttr);

    inTargetOffset = mAttr.create("targetOffset", "toff", MFnMatrixAttribute::kDouble, &status);
    INPUT_ATTR(mAttr);

    inTargetWeight = nAttr.create("targetWeight", "twght", MFnNumericData::kDouble, 1.0, &status);
    INPUT_ATTR(nAttr);

    inTargetList = cAttr.create("targetList", "tlist", &status);
    cAttr.addChild(inTargetWorldMatrix);
    cAttr.addChild(inTargetOffset);
    cAttr.addChild(inTargetWeight);
    cAttr.setArray(true);

    outConstTrans = nAttr.createPoint("constraintTranslate", "ctrans", &status);
    OUTPUT_ATTR(nAttr);

    MObject outConstRotX = uAttr.create("constraintRotateX", "crox", MFnUnitAttribute::kAngle, 0.0, &status);
    MObject outConstRotY = uAttr.create("constraintRotateY", "croy", MFnUnitAttribute::kAngle, 0.0, &status);
    MObject outConstRotZ = uAttr.create("constraintRotateZ", "croz", MFnUnitAttribute::kAngle, 0.0, &status);
    outConstRot = nAttr.create("constraintRotate", "cro", outConstRotX, outConstRotY, outConstRotZ, &status);
    OUTPUT_ATTR(nAttr);

    outConstSca = nAttr.createPoint("constraintScale", "csca", &status);
    nAttr.setDefault(1.0, 1.0, 1.0);
    OUTPUT_ATTR(nAttr);

    addAttribute(inConstraintJntOri);
    addAttribute(inConstraintRotOrder);
    addAttribute(inConstraintParInvMtx);
    addAttribute(inConstraintParSca);
    addAttribute(inTargetList);
    addAttribute(outConstTrans);
    addAttribute(outConstRot);
    addAttribute(outConstSca);
    attributeAffects(inConstraintJntOri, outConstTrans);
    attributeAffects(inConstraintRotOrder, outConstTrans);
    attributeAffects(inConstraintParInvMtx, outConstTrans);
    attributeAffects(inConstraintParSca, outConstTrans);
    attributeAffects(inTargetWorldMatrix, outConstTrans);
    attributeAffects(inTargetOffset, outConstTrans);
    attributeAffects(inTargetWeight, outConstTrans);
    attributeAffects(inConstraintJntOri, outConstRot);
    attributeAffects(inConstraintRotOrder, outConstRot);
    attributeAffects(inConstraintParInvMtx, outConstRot);
    attributeAffects(inConstraintParSca, outConstRot);
    attributeAffects(inTargetWorldMatrix, outConstRot);
    attributeAffects(inTargetOffset, outConstRot);
    attributeAffects(inTargetWeight, outConstRot);
    attributeAffects(inConstraintJntOri, outConstSca);
    attributeAffects(inConstraintRotOrder, outConstSca);
    attributeAffects(inConstraintParInvMtx, outConstSca);
    attributeAffects(inConstraintParSca, outConstSca);
    attributeAffects(inTargetWorldMatrix, outConstSca);
    attributeAffects(inTargetOffset, outConstSca);
    attributeAffects(inTargetWeight, outConstSca);

    return status;
}

MStatus ParentConstraint::compute(const MPlug& plug, MDataBlock& dataBlock){
    /*
    Node computation method:
        * plug is a connection point related to one of our node attributes (either an input or an output).
        * dataBlock contains the data on which we will base our computations.
    */
    MEulerRotation eConstJntOri = MEulerRotation(dataBlock.inputValue(inConstraintJntOri).asDouble3());
    MMatrix mConstParInv = dataBlock.inputValue(inConstraintParInvMtx).asMatrix();
    short constRotOrder = dataBlock.inputValue(inConstraintRotOrder).asShort();
    float3& constParSca = dataBlock.inputValue(inConstraintParSca).asFloat3();
    MArrayDataHandle targetListHandle = dataBlock.inputArrayValue(inTargetList);

    MMatrix mTargetsAdded = MMatrix();
    MTransformationMatrix mtxFn = MTransformationMatrix();
    double constParScaD[3] = {constParSca[0], constParSca[1], constParSca[2]};
    mtxFn.addScale(constParScaD, MSpace::kTransform);
    MMatrix mInvSca = mtxFn.asMatrix();
    MMatrix mTarget;

    for (uint32_t i = 0; i < targetListHandle.elementCount(); i++){
        targetListHandle.jumpToArrayElement(i);
        MDataHandle targetHandle = targetListHandle.inputValue();
        MMatrix mTargetW = targetHandle.child(inTargetWorldMatrix).asMatrix();
        MMatrix mOffset = targetHandle.child(inTargetOffset).asMatrix();
        double targetWeight = targetHandle.child(inTargetWeight).asDouble();

        mTarget = mOffset * (mTargetW * targetWeight);
        if (mTargetsAdded == MMatrix())
            mTargetsAdded = mTarget;
        else
            mTargetsAdded += mTarget;
    }

    MMatrix mResult = mTargetsAdded * mConstParInv * mInvSca;

    {
        MString toPrint = MString("mTargetsAdded = ");
        toPrint += mTargetsAdded[3][0];
        toPrint += ", ";
        toPrint += mTargetsAdded[3][1];
        toPrint += ", ";
        toPrint += mTargetsAdded[3][2];
        MGlobal::displayInfo(toPrint);
    }

    {
        MString toPrint = MString("mConstParInv = ");
        toPrint += mConstParInv[3][0];
        toPrint += ", ";
        toPrint += mConstParInv[3][1];
        toPrint += ", ";
        toPrint += mConstParInv[3][2];
        MGlobal::displayInfo(toPrint);
    }

    {
        MString toPrint = MString("mInvSca = ");
        toPrint += mInvSca[0][0];
        toPrint += ", ";
        toPrint += mInvSca[1][1];
        toPrint += ", ";
        toPrint += mInvSca[2][2];
        MGlobal::displayInfo(toPrint);
    }

    {
        MString toPrint = MString("mResult = ");
        toPrint += mResult[3][0];
        toPrint += ", ";
        toPrint += mResult[3][1];
        toPrint += ", ";
        toPrint += mResult[3][2];
        MGlobal::displayInfo(toPrint);
    }

    if (plug == outConstTrans){
        MDataHandle outTransHandle = dataBlock.outputValue(outConstTrans);
        MVector outTrans = MVector(mResult[3][0], mResult[3][1], mResult[3][2]);
        {
            MString toPrint = MString("outTrans = ");
            toPrint += outTrans.x;
            toPrint += ", ";
            toPrint += outTrans.y;
            toPrint += ", ";
            toPrint += outTrans.z;
            MGlobal::displayInfo(toPrint);
        }
        outTransHandle.setMFloatVector(outTrans);
        outTransHandle.setClean();
    }
    if (plug == outConstRot){
        MDataHandle outRotHandle = dataBlock.outputValue(outConstRot);
        mtxFn = MTransformationMatrix(mResult);
        MEulerRotation eRotMtx = mtxFn.eulerRotation();
        MQuaternion qRotMtx = eRotMtx.asQuaternion();
        MQuaternion qConstJntOri = eConstJntOri.asQuaternion();
        MQuaternion qOutRot = qRotMtx * qConstJntOri.invertIt();
        MEulerRotation outRot = qOutRot.asEulerRotation().reorderIt((MEulerRotation::RotationOrder)constRotOrder);
        outRotHandle.setMVector(outRot.asVector());
        outRotHandle.setClean();
    }
    if (plug == outConstSca){
        MDataHandle outScaHandle = dataBlock.outputValue(outConstSca);
        mtxFn = MTransformationMatrix(mResult);
        double sca[3];
        mtxFn.getScale(sca, MSpace::kWorld);
        MVector outSca = MVector(sca[0], sca[1], sca[2]);
        outScaHandle.setMFloatVector(outSca);
        outScaHandle.setClean();
    }

    return MStatus::kSuccess;
}